app/parsers/bulkrax/csv_parser.rb

```diff
@@ -0,0 +1,66 @@
+module Bulkrax
+  class CsvParser < ApplicationParser
+    delegate :errors, to: :importer
+
+    def self.parser_fields
+      {
+        csv_path: :string,
+        rights_statements: :string,
+        override_rights_statement: :boolean
+      }
+    end
+
+    def initialize(importer)
+      super
+    end
+
+    def records(_opts = {})
+      csv = CSV.open(
+        importer.parser_fields['csv_path'],
+        headers: true,
+        header_converters: :symbol,
+        encoding: 'utf-8'
+      )
+      csv_data = csv.read
+      raise StandardError, 'Identifier column is required' if csv_data.headers.include?(:identifier) == false
+
+      # skip rows without an identifier; remove any nil values with compact
+      csv_data.map { |row| row.to_h.compact! unless row[:identifier].blank? }.reject(&:blank?)
```

Comment:
I think this code forces all the CSV rows in to memory at the same time. That could be bad if large CSVs travel through the system.  What do you think about refactorying this method and the create_works method so that a single foreach is used? It can still have the headers and skip the bad rows (you'd skip them with `next` instead of rejecting them). For the raise you could do an `if i == 0` on your iterator (which _may_ have to be a manual counter, I'm not sure).  

The last piece that's problematic is getting the count ahead of time. I'll comment about that down there

Reply:
If you see the latest commit I've switched to having #records return the CSV.foreach enumerator and #create_works loop through that with an index. Reading around CSV.foreach is the recommended way of doing this performance-wise.

Reply:
👍 looks great